inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/scheduling/tasks/PredictionTask.java

```diff
@@ -141,42 +122,71 @@ public void run()
                                 recommender.getName(), recommender.getId(), user.getUsername(),
                                 document.getName(), document.getId(), project.getName(),
                                 project.getId());
-                        continue;
+                        continue nextRecommender;
                     }
                     
                     RecommendationEngineFactory<?> factory = recommendationService
                             .getRecommenderFactory(recommender);
                     RecommendationEngine recommendationEngine = factory.build(recommender);
 
+                    // We lazily load the CAS only at this point because that allows us to skip
+                    // loading the CAS entirely if there is no enabled layer or recommender.
+                    // If the CAS cannot be loaded, then we skip to the next document.
+                    if (!jCas.isPresent()) {
+                        try {
+                            jCas = Optional.of(documentService.readAnnotationCas(document,
+                                    user.getUsername()));
+                        }
+                        catch (IOException e) {
+                            log.error("Cannot read annotation CAS for user [{}] of document "
+                                    + "[{}]({}) in project [{}]({}) - skipping document",
+                                    user.getUsername(), document.getName(), document.getId(),
+                                    project.getName(), project.getId(), e);
+                            continue nextDocument;
+                        }
+                        try {
+                            annoService.upgradeCasIfRequired(jCas.get().getCas(), document,
+                                    user.getUsername());
+                        }
+                        catch (UIMAException | IOException e) {
+                            log.error("Cannot upgrade annotation CAS for user [{}] of document "
+                                    + "[{}]({}) in project [{}]({}) - skipping document",
+                                    user.getUsername(), document.getName(), document.getId(),
+                                    project.getName(), project.getId(), e);
+                            continue nextDocument;
+                        }
+                    }
+                    
                     try {
-                        recommendationEngine.predict(ctx, jCas.getCas());
+                        recommendationEngine.predict(ctx, jCas.get().getCas());
```

Comment:
I would `get` the Jcas once and assign it to a variable, that looks nicer

Reply:
Might defy the lazy evaluation because then I introduce a fixed evaluation point again. It might work in this case, but I'd prefer not having to worry about it. That's why I just call `get()` whenever I might need the value.